### PR TITLE
Remove unused Python variables in torch/[_-a]*

### PR DESCRIPTION
**Pull request by rec**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133492

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn @XilunWu

### CODE REVIEW

File: torch/_appdirs.py
Comment (Skylion007) [inline torch/_appdirs.py:518]: If we don't want to lose info, replaceing it `_type` etc is also valid and fine.

Comment (rec) [reply torch/_appdirs.py:518]: Dang, that's an interesting idea I didn't know, and I think I'm going to go back and do that for at least some of these variables.

Thanks for cluing me in, @Skylion007.

**Comment by rec** [comment]
This is ready to go. Because of the (very rational) limit on lines of code per code review, it ended up being six pull requests:

* `torch[_-a]*`
* `torch[b-z]*`
* `test[a-d]*`
* `test[e-z]*`
* Everything not in `torch/` or `test/`
* Turning on the flag to ruff

@amjames, can you suggest reviewers?

File: torch/_export/__init__.py
Comment (Skylion007) [inline torch/_export/__init__.py:172]: is this needed anymore?

File: torch/_functorch/_aot_autograd/collect_metadata_analysis.py
Comment (Skylion007) [inline torch/_functorch/_aot_autograd/collect_metadata_analysis.py:228]: So a lot of these are finding completely useful no-side effect functions that should be removed it appears

File: torch/_functorch/eager_transforms.py
Comment (Skylion007) [inline torch/_functorch/eager_transforms.py:1692]: I don't think tree_flatten is suppose to have side effects?

File: torch/ao/quantization/pt2e/representation/rewrite.py
Comment (Skylion007) [inline torch/ao/quantization/pt2e/representation/rewrite.py:826]: Does this have side effects? unsure

**Review by Skylion007 — COMMENTED** [review]
left some comments

File: torch/_dynamo/repro/after_dynamo.py
Comment (Skylion007) [inline torch/_dynamo/repro/after_dynamo.py:457]: Are you sure we can safely remove del args? That might have some fun memory bubble issues if `run_fwd_maybe_bwd` is expensive.

File: torch/_dynamo/variables/higher_order_ops.py
Comment (Skylion007) [inline torch/_dynamo/variables/higher_order_ops.py:2459]: technically, the behavior is slightly different as `NotImplementedError` on inspect.signature will get caught where it wasn't before.

**Review by albanD — COMMENTED** [review]
I am not sure how to handle this PR tbh
Some of these changes looks like bugfix (adding "global") and some look like just cleanup. Since there is behavior change, I do expect we should review this pretty closely.

Given that it spans so much of the codebase, it might take a while for you to get a hold onto each maintainer and get a review on this :/ 

I can try and do a pass later to see which changes require in-depth look

**Comment by rec** [comment]
Thanks for all your comments, @albanD !

There are some more minor issues later in the series, nothing earthshaking. I was looking for some maintenance task to do while waiting for more brain-twisting tasks to run.

I was weak about deleting other people's code.  I should in particular look up those tree_flattening routines to see if they possibly have side effects, which like you I doubt.

Should I be splitting these up by maintaining group? Sounds like a good idea, is this categorization still good? https://pytorch.org/docs/stable/community/persons_of_interest.html

File: torch/_inductor/codegen/cuda/gemm_template.py
Comment (albanD) [inline torch/_inductor/codegen/cuda/gemm_template.py:1378]: tbh I think it would be best for readability to change calls to these sizes below with M and N rather than removing them.

File: torch/_inductor/codegen/simd.py
Comment (albanD) [inline torch/_inductor/codegen/simd.py:1818]: split

File: torch/_inductor/codegen/wrapper.py
Comment (albanD) [inline torch/_inductor/codegen/wrapper.py:990]: This should be safe to remove I think

File: torch/_inductor/cpp_builder.py
Comment (albanD) [inline torch/_inductor/cpp_builder.py:129]: let's remove the strip/decode since we ignore their result

File: torch/_inductor/cpu_vec_isa.py
Comment (albanD) [inline torch/_inductor/cpu_vec_isa.py:122]: let's split and ask if we should check status here?

File: torch/_inductor/ir.py
Comment (albanD) [inline torch/_inductor/ir.py:1723]: let's split to make sure we want the side effect.

File: torch/_inductor/lowering.py
Comment (albanD) [inline torch/_inductor/lowering.py:5630]: No side effect here, let's delete!

File: torch/_inductor/lowering.py
Comment (albanD) [inline torch/_inductor/lowering.py:5660]: No side effect here either, we ca nremove

File: torch/_inductor/select_algorithm.py
Comment (albanD) [inline torch/_inductor/select_algorithm.py:1467]: Let's delete? Feel free to split in follow if not sure

File: torch/_library/fake_class_registry.py
Comment (albanD) [inline torch/_library/fake_class_registry.py:239]: Let's split!

File: torch/_library/fake_class_registry.py
Comment (albanD) [inline torch/_library/fake_class_registry.py:284]: I would rather we add a nice message to the assert above :)

File: torch/_numpy/testing/utils.py
Comment (albanD) [inline torch/_numpy/testing/utils.py:265]: It's weird we don't use these...
In any case, the comment above can go I guess

File: torch/_prims_common/wrappers.py
Comment (albanD) [inline torch/_prims_common/wrappers.py:370]: This is a straight up bug, let's split and add @bdhirsh to the new PR

File: torch/_subclasses/fake_impls.py
Comment (albanD) [inline torch/_subclasses/fake_impls.py:633]: looks weird, let's split

File: torch/_subclasses/fake_tensor.py
Comment (albanD) [inline torch/_subclasses/fake_tensor.py:1602]: most likely no side effect, let's split

File: torch/ao/quantization/backend_config/onednn.py
Comment (albanD) [inline torch/ao/quantization/backend_config/onednn.py:371]: This is technically changing a public API so would need to go through deprecation process.
Given that this code is most likely not maintained closely, I would just leave it as-is to avoid any issue.

File: torch/ao/quantization/backend_config/onednn.py
Comment (albanD) [inline torch/ao/quantization/backend_config/onednn.py:466]: Same as above

File: torch/ao/quantization/pt2e/representation/rewrite.py
Comment (albanD) [reply torch/ao/quantization/pt2e/representation/rewrite.py:826]: I guess it does the replace inplace, so yes

**Review by albanD — COMMENTED** [review]
No need to split everything. I left comments on the pieces that look suspicious and should be moved. A single PR for all of these should be ok as there aren't that many.
Once these are moved out, we should be able to merge this!

In case I missed any, let's put all the "# noqa: F841" ones in the separate PR as we do want careful review that they do have side effect we want to maintain!

File: torch/ao/quantization/backend_config/onednn.py
Comment (rec) [reply torch/ao/quantization/backend_config/onednn.py:371]: Urg, why did I do that?! (Changing the API) Compulsive desire to clean but out of scope! 🙂

File: torch/_dynamo/repro/after_aot.py
Comment (Skylion007) [inline torch/_dynamo/repro/after_aot.py:732]: Does this anything if the output isn't used?

File: torch/_dynamo/variables/builder.py
Comment (Skylion007) [reply torch/_dynamo/variables/builder.py:2685]: I'd prefer we use `_reason` as the other method destroys information @albanD. Better to be over-self documenting than under self-documenting.

Comment (albanD) [reply torch/_dynamo/variables/builder.py:2685]: I'm happy with preserving information! Let's update this one then!

Comment (rec) [reply torch/_dynamo/variables/builder.py:2685]: Greetings all, I am back from the Quansight retreat and raring to go on this, and am silently resolving the issues one at a time (including this one).

Thanks, @albanD for all your excellent suggestions, I should have been more aggressive about removing unused slop.

I just wanted to comment here that I didn't know about `_unused` working for `ruff` until @Skylion007 taught me about is, so after I did, I went back and fixed all the places that I'd used it, but this one (and conceivably others) missed the cut.

I'll do a review by eye of the whole thing at the end anyway.

File: torch/_dynamo/comptime.py
Comment (Skylion007) [inline torch/_dynamo/comptime.py:391]:
```suggestion
            _ctx = inner_ctx.parent()  # noqa: F841
```

File: torch/_dynamo/repro/after_dynamo.py
Comment (Skylion007) [inline torch/_dynamo/repro/after_dynamo.py:455]:
```suggestion

            del args
```

File: torch/_export/__init__.py
Comment (Skylion007) [reply torch/_export/__init__.py:172]:
```suggestion
```

File: torch/_functorch/_aot_autograd/collect_metadata_analysis.py
Comment (Skylion007) [reply torch/_functorch/_aot_autograd/collect_metadata_analysis.py:229]: This should disappear after rebase anyway, right? I thought I accepted a separate PR to remove this

**Review by Skylion007 — COMMENTED** [review]
Some more comments

File: torch/_inductor/codegen/cuda/gemm_template.py
Comment (Skylion007) [inline torch/_inductor/codegen/cuda/gemm_template.py:1379]: Hold on. Indices swapped, these two statements are not equivalent!

Comment (rec) [reply torch/_inductor/codegen/cuda/gemm_template.py:1379]: Oof, what an ineptitude on my part. I simply got rid of the unused variables. The logic around these lines of code is unnecessarily complicated but I shouldn't get outside my lane...

**Comment by rec** [comment]
@albanD @Skylion007 - all right, this is ready to go!

Sorry for the delay - there was a test that failed consistently on CI but I only managed to get to fail once on my local machine, but I found the problem.

File: torch/_inductor/bisect_helper.py
Comment (rec) [inline torch/_inductor/bisect_helper.py:471]: @albanD @Skylion007 It was this line.  Removing the local variable `cleanup` but keeping the calculation worked on my development environment but failed every time on CI!

File: torch/_inductor/codegen/cuda/gemm_template.py
Comment (albanD) [inline torch/_inductor/codegen/cuda/gemm_template.py:1380]: Wait this is still NOT fixed?

File: torch/_inductor/runtime/autotune_cache.py
Comment (albanD) [inline torch/_inductor/runtime/autotune_cache.py:255]: Let's remove this altogether as I don't think it has any side effect.

FYI @oulgen in case this is a bug

**Review by albanD — COMMENTED** [review]
Thanks for taking the time to push for this!
Only two small things and I think it should be good to go !

File: torch/_inductor/codegen/cuda/gemm_template.py
Comment (rec) [reply torch/_inductor/codegen/cuda/gemm_template.py]: That's a nasty bit of code, I cleaned it and changed the duplicated condition to actually make sense, let's see what the CI says!

File: torch/_inductor/runtime/autotune_cache.py
Comment (rec) [reply torch/_inductor/runtime/autotune_cache.py:255]: Done.

**Comment by rec** [comment]
@albanD - done!

**Review by albanD — APPROVED** [review]
Thanks!

**Comment by rec** [comment]
@cyyever, @clee2000 @albanD

A rebase isn't going to work, because the commit is already contained in `main`!  The rebase process sees the identical commit, just removes it, and then the revert comes after that, and we're back where we started. 😁 

The solution I used was to go to the desired commit ID - I picked `main` - and then use `git cherry-pick` to textually apply the commit again.

If `upstream` is the pytorch/pytorch repo, these commands would do it:

```
git fetch upstream main
git reset --hard upstream/main
git cherry-pick upstream/gh/rec/45/head
ghstack
```

I did something like this, and it's cooking right now.

**Comment by cyyever** [comment]
Fine , merge it!

**Pull request closed without merging**
